sparse/_dok.py

```diff
@@ -308,15 +309,54 @@ def __getitem__(self, key):
             return self.fill_value
 
     def __setitem__(self, key, value):
-        key = normalize_index(key, self.shape)
-        value = np.asanyarray(value)
+        value = np.asanyarray(value, dtype=self.dtype)
```

Comment:
```suggestion
        value = np.asarray(value, dtype=self.dtype)
```
`np.asanyarray` lets subclasses through. There are prominent subclasses of `np.ndarray` that don't follow the [Liskov Substitution Principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle), and are therefore not fit for consumption by this function.